controller/linodefirewall_controller.go

```diff
@@ -0,0 +1,273 @@
+
+		switch {
+		case errors.Is(err, errTooManyIPs):
+			// Cannot reconcile firewall with too many ips, wait for an update to the spec
+			return ctrl.Result{}, nil
+		case util.IsRetryableError(err) && !reconciler.HasConditionSeverity(fwScope.LinodeFirewall, clusterv1.ReadyCondition, clusterv1.ConditionSeverityError):
+			logger.Info(fmt.Sprintf("re-queuing Firewall %s", action))
+
+			return ctrl.Result{RequeueAfter: reconciler.DefaultFWControllerReconcilerDelay}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	if action == "create" && fwScope.LinodeFirewall.Spec.FirewallID != nil {
+		r.Recorder.Event(fwScope.LinodeFirewall, corev1.EventTypeNormal, "Created", fmt.Sprintf("Created Firewall %d", *fwScope.LinodeFirewall.Spec.FirewallID))
+	} else {
+		r.Recorder.Event(fwScope.LinodeFirewall, corev1.EventTypeNormal, "Updated", fmt.Sprintf("Updated Firewall %d", *fwScope.LinodeFirewall.Spec.FirewallID))
+	}
+	fwScope.LinodeFirewall.Status.Ready = true
+
+	return ctrl.Result{}, nil
+}
+
+func (r *LinodeFirewallReconciler) reconcileDelete(
+	ctx context.Context,
+	logger logr.Logger,
+	fwScope *scope.FirewallScope,
+) (ctrl.Result, error) {
+	logger.Info("deleting firewall")
+
+	if fwScope.LinodeFirewall.Spec.FirewallID == nil {
+		logger.Info("firewall ID is missing, nothing to do")
+		controllerutil.RemoveFinalizer(fwScope.LinodeFirewall, infrav1alpha2.FirewallFinalizer)
+
+		return ctrl.Result{}, nil
+	}
+
+	err := fwScope.LinodeClient.DeleteFirewall(ctx, *fwScope.LinodeFirewall.Spec.FirewallID)
+	if util.IgnoreLinodeAPIError(err, http.StatusNotFound) != nil {
+		logger.Error(err, "failed to delete Firewall")
+
+		if fwScope.LinodeFirewall.ObjectMeta.DeletionTimestamp.Add(reconciler.DefaultTimeout(r.ReconcileTimeout, reconciler.DefaultFWControllerReconcileTimeout)).After(time.Now()) {
+			logger.Info("re-queuing Firewall deletion")
+
+			return ctrl.Result{RequeueAfter: reconciler.DefaultFWControllerReconcilerDelay}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	controllerutil.RemoveFinalizer(fwScope.LinodeFirewall, infrav1alpha2.FirewallFinalizer)
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LinodeFirewallReconciler) SetupWithManager(mgr ctrl.Manager, options crcontroller.Options) error {
+	linodeFirewallMapper, err := kutil.ClusterToTypedObjectsMapper(
+		r.TracedClient(),
+		&infrav1alpha2.LinodeFirewallList{},
+		mgr.GetScheme(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create mapper for LinodeFirewalls: %w", err)
+	}
+	err = ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1alpha2.LinodeFirewall{}).
+		WithOptions(options).
+		WithEventFilter(
+			predicate.And(
+				predicates.ResourceNotPausedAndHasFilterLabel(mgr.GetLogger(), r.WatchFilterValue),
+				predicate.GenerationChangedPredicate{},
+				predicate.Funcs{DeleteFunc: func(e event.DeleteEvent) bool { return false }},
```

Comment:
I think we've landed on this predicate no longer being necessary?

Reply:
yep